src/protocol/vm/tycho_decoder.rs

```diff
@@ -0,0 +1,251 @@
+use std::{
+    collections::HashMap,
+    time::{SystemTime, UNIX_EPOCH},
+};
+
+use ethers::types::{H160, H256, U256};
+
+use tycho_client::feed::{synchronizer::ComponentWithState, Header};
+
+use crate::{
+    evm::{simulation_db::BlockHeader, tycho_db::PreCachedDB},
+    protocol::{errors::InvalidSnapshotError, vm::state::VMPoolState, BytesConvertible},
+};
+
+#[allow(dead_code)]
+trait TryFromWithBlock<T> {
+    type Error;
+    async fn try_from_with_block(value: T, block: Header) -> Result<Self, Self::Error>
+    where
+        Self: Sized;
+}
+
+impl From<Header> for BlockHeader {
+    fn from(header: Header) -> Self {
+        let now = SystemTime::now()
+            .duration_since(UNIX_EPOCH)
+            .expect("Time went backwards")
```

Comment:
time travelling is possible ok?

Reply:
Exactly.... that's why I have it as an expectation here 👀